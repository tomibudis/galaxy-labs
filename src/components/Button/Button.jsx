import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import { THEME, SIZE } from '#constants';

export const TESTID = {
  BTN: 'btnTestID',
  LOADING: 'btnLoadingTestID',
};

const Button = ({ isLoading, theme, children, size, ...restProps }) => {
  const getTheme = useMemo(() => {
    switch (theme) {
      case THEME.PRIMARY:
        return 'btn-primary';
      case THEME.SECONDARY:
        return 'btn-secondary';
      case THEME.WARNING:
        return 'btn-warning';
      case THEME.DANGER:
        return 'btn-danger text-white';
      default:
        return 'btn-primary';
    }
  }, [theme]);

  const getSize = useMemo(() => {
    switch (size) {
      case SIZE.SM:
        return 'btn-sm';
      case SIZE.MD:
        return 'btn-md';
      case SIZE.LG:
        return 'btn-lg';
      default:
        return 'btn-md';
    }
  }, [size]);

  return (
    <button
      type="button"
      disabled={isLoading}
      className={cx('btn', getTheme, getSize)}
      data-testid={TESTID.BTN}
      {...restProps}
    >
      {isLoading && (
        <span
          className="spinner-border spinner-border-sm me-2"
          role="status"
          aria-hidden="true"
          data-testid={TESTID.LOADING}
        />
      )}
      <span className="ml-3">{children}</span>
    </button>
  );
};

Button.propTypes = {
  /** Set loading button [true | false] */
  isLoading: PropTypes.bool,
  /** Set size of button */
  size: PropTypes.oneOf(['sm', 'md', 'lg']),
  /** Set theme of button [string] */
  theme: PropTypes.oneOf(['primary', 'secondary', 'warning', 'danger']),
  /** Children node */
  children: PropTypes.node,
};

Button.defaultProps = {
  size: 'md',
  theme: 'primary',
};

export default Button;
