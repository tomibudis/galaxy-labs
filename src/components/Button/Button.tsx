import React, { useMemo } from 'react';

import PropTypes from 'prop-types';
import cx from 'classnames';

import { THEME, SIZE } from '#constants/index';
import { ButtonProps } from '#interface/index';

export const TESTID = {
  BTN: 'btnTestID',
  LOADING: 'btnLoadingTestID',
};

const Button: React.FC<ButtonProps> = ({
  isLoading,
  variant,
  children,
  size,
  ...restProps
}): JSX.Element => {
  const getTheme: string = useMemo(() => {
    switch (variant) {
      case THEME.PRIMARY:
        return 'btn-primary';
      case THEME.SECONDARY:
        return 'btn-secondary';
      case THEME.WARNING:
        return 'btn-warning';
      case THEME.DANGER:
        return 'btn-danger text-white';
      default:
        return 'btn-primary';
    }
  }, [variant]);

  const getSize: string = useMemo(() => {
    switch (size) {
      case SIZE.SM:
        return 'btn-sm';
      case SIZE.MD:
        return 'btn-md';
      case SIZE.LG:
        return 'btn-lg';
      default:
        return 'btn-md';
    }
  }, [size]);

  return (
    <button
      type="button"
      disabled={isLoading}
      className={cx('btn', getTheme, getSize)}
      data-testid={TESTID.BTN}
      {...restProps}
    >
      {isLoading && (
        <span
          className="spinner-border spinner-border-sm me-2"
          role="status"
          aria-hidden="true"
          data-testid={TESTID.LOADING}
        />
      )}
      <span className="ml-3">{children}</span>
    </button>
  );
};

Button.propTypes = {
  /** Set loading button [true | false] */
  isLoading: PropTypes.bool,
  /** Set size of button */
  size: PropTypes.oneOf(['sm', 'md', 'lg']),
  /** Set variant of button [string] */
  variant: PropTypes.oneOf(['primary', 'secondary', 'warning', 'danger']),
  /** Children node */
  children: PropTypes.node,
};

Button.defaultProps = {
  size: 'md',
  variant: 'primary',
  isLoading: false,
};

export default Button;
