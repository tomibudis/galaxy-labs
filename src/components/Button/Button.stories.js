import React from 'react';
import Button from './Button';

export default {
  title: 'Components/Button',
  component: Button,
  argTypes: {
    children: {
      table: {
        disable: true,
      },
    },
  },
};

const Template = (args) => {
  return <Button {...args}>{args.theme}</Button>;
};

export const TypeButton = Template.bind({});
TypeButton.parameters = {
  controls: { exclude: ['isLoading'] },
};
TypeButton.args = {
  theme: 'primary',
};
