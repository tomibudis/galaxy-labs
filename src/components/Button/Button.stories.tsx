import React from 'react';

import { Meta, ComponentStory } from '@storybook/react';

import Button from './Button';

export default {
  title: 'Components/Button',
  component: Button,
  argTypes: {
    children: {
      table: {
        disable: true,
      },
    },
  },
} as Meta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => {
  return <Button {...args}>{args.variant}</Button>;
};

export const TypeButton = Template.bind({});
TypeButton.args = {
  variant: 'primary',
  isLoading: false,
};
