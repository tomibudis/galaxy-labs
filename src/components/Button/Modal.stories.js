import React from 'react';

import Button from './Button';

export default {
  title: 'Components/Button',
  component: Button,
};

const Template = (args) => <Button {...args} />;

export const ButtonPrimary = Template.bind({});
ButtonPrimary.args = {};

