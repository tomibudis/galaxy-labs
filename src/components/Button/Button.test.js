import React from 'react';
import { act, cleanup, fireEvent, render } from '@testing-library/react';

import Button, { TESTID } from './Button';

afterEach(() => {
  cleanup();
  mockOnClick.mockClear();
});

const mockOnClick = jest.fn();

const ButtonComponent = (props) => {
  return <Button {...props}>{props?.children || 'Test Component'}</Button>;
};

describe('<Button />', () => {
  it('render component', () => {
    const { getByTestId } = render(<ButtonComponent />);
    expect(getByTestId(TESTID.BTN)).toBeTruthy();
  });

  it('triggered onClick', () => {
    const { getByTestId } = render(<ButtonComponent onClick={mockOnClick} />);
    act(() => {
      fireEvent.click(getByTestId(TESTID.BTN));
    });
    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });

  test.each([
    ['primary', '.btn-primary'],
    ['secondary', '.btn-secondary'],
    ['warning', '.btn-warning'],
    ['danger', '.btn-danger'],
  ])('render %s button', (theme, expected) => {
    const { baseElement } = render(<Button theme={theme} />);
    expect(baseElement.querySelector(expected)).toBeTruthy();
  });

  test.each([
    ['sm', '.btn-sm'],
    ['md', '.btn-md'],
    ['lg', '.btn-lg'],
  ])('render size %s button', (size, expected) => {
    const { baseElement } = render(<Button size={size} />);
    expect(baseElement.querySelector(expected)).toBeTruthy();
  });

  it('matched with snapshot file', () => {
    const { baseElement } = render(<Button />);
    expect(baseElement).toMatchSnapshot();
  });
});
