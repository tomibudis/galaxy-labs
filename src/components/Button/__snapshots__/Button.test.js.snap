// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> matched with snapshot file 1`] = `
<body>
  <div>
    <button
      class="btn btn-primary btn-md"
      data-testid="btnTestID"
      type="button"
    >
      <span
        class="ml-3"
      />
    </button>
  </div>
</body>
`;
